### PR TITLE
Add recommendations around language injection

### PR DESCRIPTION
**Pull request by ksobolew**
This is an enhancement of the section about IntelliJ IDEA's inspections. Proper handling of language injection is unfortunately not automatic, and there is no objective truth to how to do it. But since we're working with SQL statements a lot, I find it useful.

The text is a result of a very not-wide-ranging discussion we had about this subject. It may be seen as a kind of imposition on our part, so please share your opinion on this. Is there anything else that we can do? Is there a better way that I don't know about? Is it too wordy? Is it necessary at all?

Thanks!

### CODE REVIEW

**Comment by findepi** [comment]
I think IntelliJ CE doesn't support SQL language injection out of the box.
We should not make it seem that Ultimate edition is required to work with the codebase.

**Comment by ksobolew** [comment]
> I think IntelliJ CE doesn't support SQL language injection out of the box.
> We should not make it seem that Ultimate edition is required to work with the codebase.

Oh. That is a good point I haven't even considered

**Comment by skrzypo987** [comment]
> I think IntelliJ CE doesn't support SQL language injection out of the box.
> We should not make it seem that Ultimate edition is required to work with the codebase.

IMO Language annotations in APIs are an added value regardless of the IDE. There are used in many classes right now and formalizing it in docs seems a good idea.

As of language injection warnings I fully agree, we should not assume that the editor has IDEA Ultimate

**Comment by ksobolew** [comment]
> IMO Language annotations in APIs are an added value regardless of the IDE. There are used in many classes right now and formalizing it in docs seems a good idea.
> 
> As of language injection warnings I fully agree, we should not assume that the editor has IDEA Ultimate

OK, I will try to reword it in a way that separates these two concerns.

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -119,6 +119,14 @@ Disable the following inspections:
 - ``Java | Performance | Call to 'Arrays.asList()' with too few arguments``,
```
Comment (kokosing) [inline DEVELOPMENT.md]: Commit message:
 - typo `Byt`
 - `Byt since we're working with SQL statements a lot, I find it useful.` maybe let's remove this statement. It does not add much value to the commit.

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -119,6 +119,14 @@ Disable the following inspections:
...
+- Disable inspection ``SQL | No data source configured``: Language injection for SQL is made significantly more useful when it is also aware of the structure of the database; unfortunately in Trino we don't have a single data source - and the tests may even be using a completely made-up database structure - so this inspection is not very helpful.
+- Inspection ``Language injection | Language mismatch``: This inspection will highlight any local variable which is used in a context annotated with language injection, but is not annotated itself. We don't recommend disabling it outright, because many people find it useful, but we're also not keen on over-using the annotations where the language can be perfectly well inferred from context (e.g. when the variable is used as an argument for a parameter annotated with specific language).
+
+Considering the above, we recommend that you annotate all API parameters which are expecting to take a `String` containing an SQL statement (or any other language, like regular expressions). You may annotate also any local variables in new code, but refrain from doing it in existing code, as it only generates noise in the commit history.
+
```
Comment (kokosing) [inline DEVELOPMENT.md]: This is very verbose. In order to understand what it is all about a reader is required to spend noticeable amount of energy.  This concerns me as I am afraid people won't read it.

How about:
```
### Language injection

To use language injection certain code elements can be annotated with the 
`@org.intellij.lang.annotations.Language` annotation. 

To make it useful we recommend:
- Set ``Languages & Frameworks | SQL Dialects``
- Disable inspection ``SQL | No data source configured``:
- Disable inspection ``Language injection | Language mismatch``.

Considering the above, we recommend to annotate with `@Language`:
 -  all API parameters which are expecting to take a `String` 
 containing an SQL statement (or any other language, like regular expressions). 
  - local variables which otherwise would no be properly recognized 
  by IDE  as language injection.
```

**Comment by kokosing** [comment]
@trinodb/maintainers Any opininions?

**Comment by findepi** [comment]
> @trinodb/maintainers Any opinions?

see https://github.com/trinodb/trino/pull/7536#issuecomment-816620445

i think we could put this into a section "do this if you have Intellij Ultimate", separate from other sections of the DEV doc, unless there is a way to make SQL language injection to work with IntelliJ CE.
Then, such instructions should be added as well.

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -124,6 +124,21 @@ Enable errorprone ([Error Prone Installation#IDEA](https://errorprone.info/docs/
...
 
+### Language injection in IDE
+
+Because we work a lot with SQL, in many places you'll encounter elements annotated with the `@org.intellij.lang.annotations.Language` annotation, like in: `@Language("SQL")`. It serves two purposes:
+
+- As a documentation: when a method parameter is annotated with a language annotation, the callers are aware that the argument should be a value conforming to the specific language;
+- As an aid in static analysis: when the IDE supports it, it may use this information to highlight mis-using code, and provide things like language injection in conforming code.
+
+For this reason we recommend that you annotate all API parameters which are expecting to take a `String` containing an SQL statement (or any other language, like regular expressions). You may also annotate local variables in new code, but refrain from doing it in existing code, as it only generates noise in the commit history.
```
Comment (losipiuk) [inline DEVELOPMENT.md:134]: Why `refrain from doing it in existing code`. IMO if we decide to annotate, we should also move towards consistency all over the codebase.

File: DEVELOPMENT.md (hunk 2 above)
Comment (losipiuk) [reply DEVELOPMENT.md]: I would not worry too much about people not reading it. I would expect that if someone reads DEVELOPMENT.md at all, they will over go through this (not that I expect that many people would any part of that file :) ).

Still I like the shorter version

File: DEVELOPMENT.md (hunk 3 above)
Comment (skrzypo987) [reply DEVELOPMENT.md:134]: We had an offline discussion about annotating variables and decided that there is little added value and a lot of boilerplate code, so the decision is up to the creator.
Maybe the best idea is to skip this last phrase.

Comment (findepi) [reply DEVELOPMENT.md:134]: > ... but refrain from doing it in existing code ...

this is generally applicable to any refactor, nothing special here

unless you want to discourage someone from sending a PR  that just adds `@Language("SQL")` because the low-value added doesn't motivate spending review time on these? (which i could see being _possible_)

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -124,6 +124,21 @@ Enable errorprone ([Error Prone Installation#IDEA](https://errorprone.info/docs/
...
+
+Language injection is a feature e.g. in Intellij which provides additional syntax highlighting, code completion etc. inside strings, which it would otherwise treat as opaque blobs of characters. Before it's actually useful, though, there are three configurations that you may have to change:
```
Comment (findepi) [inline DEVELOPMENT.md:136]: > e.g. in Intellij

the annotation is `@org.intellij` so maybe `e.g.` is not needed here

File: DEVELOPMENT.md (hunk 3 above)
Comment (kokosing) [reply DEVELOPMENT.md:134]: I would remove: `but refrain from doing it in existing code, as it only generates noise in the commit history.`

Instead I would say: `You may also annotate local variables in new code, if it helps IDE to recognize the language injection`. So I would not recommend to add for case where IDE already recognized that (possibly because variable was used in the method call).

**Comment by ksobolew** [comment]
Pushed a new version based on the one proposed by @kokosing.

**Review by losipiuk — APPROVED** [review]

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -124,6 +124,19 @@ Enable errorprone ([Error Prone Installation#IDEA](https://errorprone.info/docs/
 - In ``Java Compiler`` tab, select ``Javac with error-prone`` as the compiler,
 - Update ``Additional command line parameters`` with ``-XepExcludedPaths:.*/target/generated-(|test-)sources/.* -XepDisableAllChecks -Xep:MissingOverride:ERROR ......`` (for current recommended list of command line parameters, see the top level ``pom.xml``, the definition of the ``errorprone-compiler`` profile.
 
+### Language injection in IDE
+
+In order to enable language injection inside Intellij IDEA, some code elements can be annotated with the `@org.intellij.lang.annotations.Language` annotation. To make it useful, we recommend:
+
+- Set the project-wide SQL dialect in ``Languages & Frameworks | SQL Dialects`` - "Generic SQL" is a decent choice here;
```
Comment (kokosing) [inline DEVELOPMENT.md:131]: Above we are using `,` instead of `;`. Should we update here or there?

**Review by kokosing — APPROVED** [review]
% nit

File: DEVELOPMENT.md (hunk 5 above)
Comment (ksobolew) [reply DEVELOPMENT.md]: Indeed; I changed them to `,` for consistency.

**Comment by kokosing** [comment]
I will leave it open for a while to leave a chance for others to chime in, there is no rush. If there will be no more comments please ping on Monday so I will merge it.

File: DEVELOPMENT.md [hunk 6]
```diff
@@ -124,6 +124,19 @@ Enable errorprone ([Error Prone Installation#IDEA](https://errorprone.info/docs/
...
+In order to enable language injection inside Intellij IDEA, some code elements can be annotated with the `@org.intellij.lang.annotations.Language` annotation. To make it useful, we recommend:
+
+- Set the project-wide SQL dialect in ``Languages & Frameworks | SQL Dialects`` - "Generic SQL" is a decent choice here,
+- Disable inspection ``SQL | No data source configured``,
+- Optionally disable inspection ``Language injection | Language mismatch``.
+
+Even if the IDE does not support language injection this annotation is useful for documenting the API's intent. Considering the above, we recommend annotating with `@Language`:
+
+- All API parameters which are expecting to take a `String` containing an SQL statement (or any other language, like regular expressions),
+- Local variables which otherwise would not be properly recognized by IDE for language injection.
```
Comment (skrzypo987) [inline DEVELOPMENT.md:138]: 1."which otherwise would not be properly recognized by IDE"
Just thinking logically, how a person not having InteliJ Ultimate knows what that means?

2.So we want to make annotations in API and annotation before every local variable an equally recommended thing. The current state of the code does not represent this - there is a lot of args with this annotation and almost no variables.

Comment (kokosing) [reply DEVELOPMENT.md:138]: >  how a person not having InteliJ Ultimate knows what that means

They don't know, so they don't care, so they don't change.

> So we want to make annotations in API and annotation before every local variable an equally recommended thing.

No. Only in cases where `Local variables which otherwise would not be properly recognized by IDE for language injection.` otherwise annotation should not be added. I expect `a lot of args with this annotation and almost no variables.`

Comment (skrzypo987) [reply DEVELOPMENT.md:138]: > They don't know, so they don't care, so they don't change.
That sounds terrible, but I guess you are right. I don't understand it and already started not caring.
